### ISSUE TITLE
"Hot Swap" configs based on MP or SP menus?

### ISSUE THREAD

**Issue by zagatozee** [mdias/rs_asio#223]
TL:DR version, can the way the RS_ASIO hackery is done, handle having a configuration change made after it has initialised at start up?

Hey there. I'm one of the people that did some of the work on RSMods.
I recently got myself a new ASIO interface that has 2 input channels.

From RSMods, we know how to detect the menu the game is in, so I was wondering; if we added a detection for the multiplayer menus being active, could RS_ASIO handle being told to change to an alternate config file, one with both input channels active, or would the way the ASIO hackery is done, fall on it's face trying to change the configs after the title has loaded?

I'm picturing the RS_ASIO.ini file having - instead of the input 1 / input 2 / mic definitions as currently, changing it so Player 1 and Player 2 inputs are defined, and having the hijacked dll part only activate the player 2 input when the MP menu has been detected as active.

If you have any thoughts on this before I go diving into it too deep, I'd love to hear them. Rocksmith is infamously stupid about handling having multiple "RTC" connected, and if we can improve the Quality of Life experience on that one with ASIO gear, I'd love to spend some time making it possible.

**Comment by zagatozee** [comment]
FYI, over at RSMods, we have a Proof Of Concept working for this issue. 

https://github.com/Lovrom8/RSMods/commit/5c4071b56a2ed18041b8b412ae460b9593355bbe

We could implement it into RSMods and probably will make it a toggleable option for the rare use case where it is worth having in a non RS_ASIO setup, but I at least, feel it would be better suited being built directly into RS_ASIO.

I've tested it on ASIO and 2 x RTC setups, and it works.

Only issue, is if Windows decides that the most recently connected cable (2 x RTC setup) has to be player 1, you can only swap this via the Mulitplayer menu, and need to on every startup of the game.

 We have a thought for how to make that easier to live with - but from an RS_ASIO point of view, it shouldn't be an issue as in our testing, the player 1 input as defined in the RS_ASIO.ini is always the player 1 input on startup.

Link above is the Proof Of Concept, however the magician doing it wants to improve the spell a little still, but POC is good.

**Comment by zagatozee** [comment]
Just for clarity, the above method isn't "fixing" the issue by changing the configs as needed, it is just telling RS to shush with the errors and get on with it when in a Single Player mode. Unorthodox maybe, but it works.

**Comment by ffio1** [comment]
I cleaned up the patch, as the patch that ZZ linked was definitely a PoC and had a couple bugs. This version should be more stable, and doesn't rely on menu names / multiplayer status as we just jump over the code that throws up the messagebox. https://github.com/Lovrom8/RSMods/commit/9b835dac4f7dd48bddd7bcdd53f4ffa49d5f847d (disregard the multiplayer checks in DllMain, those were removed in https://github.com/Lovrom8/RSMods/commit/5ae963c70414047cb344ee9978de5433803b9626 as they were 
causing the menu to pop up if the user left multiplayer right after opening it).

It should be fine to apply this patch without needing to put the old bytes back. As ZZ said, it would be ideal for RS_ASIO to implement it but if it isn't practical we can always add it as a mod to RSMods.

Feel free to reach out if you have any questions, and I can try to answer them.

**Comment by ffio1** [comment]
Later today, I'll see if I can make a build of RS_ASIO with the same patch applied.

**Comment by mdias** [comment]
I'm closing this since the patch was merged.
Thank you again for the contribution :)
